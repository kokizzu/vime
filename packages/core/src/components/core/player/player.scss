vime-player {
  /**
   * @prop --vime-font-family: A custom font family to be used throughout the player.
   * 
   * @prop --vime-theme: A custom theme (color) to be used throughout the player. Any valid CSS 
   * `color` property should work (HEX, RGBA, HLS, ...).
   */

  --vime-font-family: 'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif;

  box-sizing: border-box;
  direction: ltr;
  font-family: var(--vime-font-family);
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: subpixel-antialiased;
  -webkit-tap-highlight-color: transparent;
  font-variant-numeric: tabular-nums;
  font-weight: 500;
  line-height: 1.7;
  width: 100%;
  display: block;
  max-width: 100%;
  min-width: 414px;
  position: relative;
  text-shadow: none;
  outline: 0;
  transition: box-shadow 0.3s ease;

  *,
  *::after,
  *::before {
    box-sizing: inherit;
  }

  button {
    font: inherit;
    line-height: inherit;
    width: auto;

    &::-moz-focus-inner {
      border:0;
    }
  }

  // Fix 300ms delay
  a,
  button,
  input,
  label {
    touch-action: manipulation;
  }

  &.video {
    height: 0;
    background: #000;
    overflow: hidden;
  }

  &.fullscreen {
    background: #000;
    margin: 0;
    border-radius: 0;
    width: 100%;
    height: 100%;
    padding-bottom: 0 !important;
  }

  .blocker {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: $z-index-blocker;
    display: inline-block;
  }
}
