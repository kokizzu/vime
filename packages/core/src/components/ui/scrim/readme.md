# vime-scrim

A darkened overlay that covers the current video to enable content placed on it to be more visible.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <vime-scrim active></vime-scrim>
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                          | Type      | Default |
| -------- | --------- | ------------------------------------ | --------- | ------- |
| `active` | `active`  | Whether the scrim is visible or not. | `boolean` | `false` |


## Events

| Event      | Description                            | Type                |
| ---------- | -------------------------------------- | ------------------- |
| `willHide` | Emitted when the scrim will be hidden. | `CustomEvent<void>` |
| `willShow` | Emitted when the scrim will be shown.  | `CustomEvent<void>` |


## CSS Custom Properties

| Name         | Description                        |
| ------------ | ---------------------------------- |
| `--scrim-bg` | The background color of the scrim. |


## Dependencies

### Used by

 - [vime-default-controls](../controls/default-controls)

### Graph
```mermaid
graph TD;
  vime-default-controls --> vime-scrim
  style vime-scrim fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
