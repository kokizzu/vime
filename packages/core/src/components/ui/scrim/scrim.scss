vime-scrim {
  /**
   * @prop --scrim-bg: The background color of the scrim.
   */

  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: $z-index-scrim;
  background: var(--scrim-bg);
  display: inline-block;
  pointer-events: none;
  @include invisible();
  
  &.hidden {
    display: none;
  }

  &.active {
    @include visible();
  }
}