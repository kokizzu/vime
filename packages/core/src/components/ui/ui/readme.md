# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
