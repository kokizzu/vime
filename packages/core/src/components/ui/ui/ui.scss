vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}