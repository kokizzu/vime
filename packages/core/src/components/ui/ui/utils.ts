import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
