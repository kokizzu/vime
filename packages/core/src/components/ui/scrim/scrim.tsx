import {
  h, Component, Host, State, Watch, Prop, Event, EventEmitter,
} from '@stencil/core';
import { PlayerProps, PlayerProp } from '../../core/player/PlayerProp';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';

@Component({
  tag: 'vime-scrim',
  styleUrl: 'scrim.scss',
})
export class Scrim {
  @State() isHidden = true;

  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  @Watch('isVideoView')
  onVideoViewChange() {
    this.isHidden = !this.isVideoView;
  }

  /**
   * Whether the scrim is visible or not.
   */
  @Prop() active = false;

  /**
   * Emitted when the scrim will be shown.
   */
  @Event({ bubbles: false }) willShow!: EventEmitter<void>;

  /**
   * Emitted when the scrim will be hidden.
   */
  @Event({ bubbles: false }) willHide!: EventEmitter<void>;

  @Watch('active')
  @Watch('isHidden')
  onVisibilityChange() {
    (!this.isHidden && this.active) ? this.willShow.emit() : this.willHide.emit();
  }

  render() {
    return (
      <Host
        class={{
          hidden: this.isHidden,
          active: this.active,
        }}
      >
        <div />
      </Host>
    );
  }
}

openPlayerWormhole(Scrim, [PlayerProp.IsVideoView]);
