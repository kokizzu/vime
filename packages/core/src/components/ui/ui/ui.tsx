import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        {canShowCustomUI && <slot />}
      </Host>
    );
  }
}

openPlayerWormhole(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
