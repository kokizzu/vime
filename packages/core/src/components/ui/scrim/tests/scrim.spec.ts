import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../ui/tests';
import { PlayerProp } from '../../../core/player/PlayerProp';
import { ViewType } from '../../../core/player/ViewType';
import { Scrim } from '../scrim';

let page: SpecPage;
let provider: HTMLVimeFaketubeElement;
let scrim: HTMLVimeScrimElement;

beforeEach(async () => {
  ({ page, provider } = await newUISpecPage(
    [Scrim],
    '<vime-scrim />',
  ));

  scrim = page.root!.querySelector('vime-scrim')!;
});

it('should be structurally sound', () => {
  expect(scrim).toMatchSnapshot();
});

it('should not render if not a video view', async () => {
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Audio);
  await page.waitForChanges();
  expect(scrim).not.toHaveClass('enabled');
});

it('should render if a video view', async () => {
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Video);
  await page.waitForChanges();
  expect(scrim).toHaveClass('enabled');
});

it('should be visible if active', async () => {
  scrim.active = true;
  await page.waitForChanges();
  expect(scrim).toHaveClass('active');
});

it('should not be visible if not active', async () => {
  scrim.active = false;
  await page.waitForChanges();
  expect(scrim).not.toHaveClass('active');
});

it('should emit willShow event when visible', async () => {
  const cb = jest.fn();
  scrim.addEventListener('willShow', cb);
  scrim.active = true;
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Video);
  await page.waitForChanges();
  expect(cb).toHaveBeenCalled();
});

it('should emit willHide event when not visible', async () => {
  const cb = jest.fn();
  scrim.addEventListener('willHide', cb);
  scrim.active = true;
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Video);
  await page.waitForChanges();
  scrim.active = false;
  await page.waitForChanges();
  expect(cb).toHaveBeenCalled();
});
