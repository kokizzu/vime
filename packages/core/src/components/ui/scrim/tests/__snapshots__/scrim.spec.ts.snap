// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-scrim class="hidden">
  <div></div>
</vime-scrim>
`;
